cmd/vsphere-xcopy-volume-populator/internal/powerstore/powerstore.go

```diff
@@ -0,0 +1,297 @@
+package powerstore
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/dell/gopowerstore"
+	"github.com/kubev2v/forklift/cmd/vsphere-xcopy-volume-populator/internal/populator"
+	"k8s.io/klog/v2"
+)
+
+type PowerstoreClonner struct {
+	Client gopowerstore.Client
+}
+
+// CurrentMappedGroups implements populator.StorageApi.
+func (p *PowerstoreClonner) CurrentMappedGroups(targetLUN populator.LUN, mappingContext populator.MappingContext) ([]string, error) {
+	if targetLUN.IQN == "" {
+		return nil, fmt.Errorf("target LUN IQN is required")
+	}
+
+	ctx := context.Background()
+	mappings, err := p.Client.GetHostVolumeMappingByVolumeID(ctx, targetLUN.IQN)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get volume mappings for %s: %w", targetLUN.Name, err)
+	}
+
+	mappedHosts := make([]string, 0, len(mappings))
+	for _, mapping := range mappings {
+		host, err := p.Client.GetHost(ctx, mapping.HostID)
+		if err != nil {
+			klog.Warningf("Failed to get host info for host ID %s: %s", mapping.HostID, err)
+			continue
+		}
+		mappedHosts = append(mappedHosts, host.Name)
+	}
+
+	klog.Infof("Volume %s is currently mapped to hosts: %v", targetLUN.Name, mappedHosts)
+	hostName := mappingContext["hostName"].(string)
+	initiatorGroup := mappingContext["initiatorGroup"].(string)
+	if slices.Contains(mappedHosts, hostName) {
+		mappedHosts = append(mappedHosts, initiatorGroup)
+	}
+	return mappedHosts, nil
+}
+
+// EnsureClonnerIgroup implements populator.StorageApi.
+func (p *PowerstoreClonner) EnsureClonnerIgroup(initiatorGroup string, adapterIds []string) (populator.MappingContext, error) {
+	klog.Infof("ensuring host %s for adapters %v", initiatorGroup, adapterIds)
+
+	ctx := context.Background()
+	mappingContext := make(map[string]any)
```

Comment:
good practice. In the near future I plan to refactor the mapping context to a struct with the map internlized,so we always have a zero value, so it is harder to panic